### PR TITLE
Add RegularizedIterativeSENSEReconstruction

### LINKED ISSUE #6: l2_tikonov_solver
Image

Check list:
- [ ]     Create branch
- [ ]     Implement code
- [ ]     Add comments
- [ ]     Implement tests
- [ ]     Squash commmits
- [ ]     Create PR
- [ ]     Assign reviewer(s) to PR

### LINKED ISSUE #321: iterative Sense + Tikonov regularisation
supersedes #6 

Design suggestion for IterativeSENSE reconstruction algorithm
```
class iterativeSENSE(Reconstruction):
     acquisition_model: LinearOperator
     initial_value: None | IData
     n_max_iterations: int
     dcf: DcfOperator | None
```

Design suggestion for IterativeSENSEWithTikonov reconstruction algorithm
```
class iterativeSENSEWithTikonov(Reconstruction):
     acquisition_model: LinearOperator
     initial_value: None | IData
     n_max_iterations: int
     dcf: DcfOperator | None
     regularisation_image: None | IData
     regularisation_weight: float
```
If regularisation_weight == 0 or regularisation_image is None this defaults to `iterativeSENSE`

Open questions:

- Should we have an additional l2_regularised_least_squares optimizer which solves the more general problem `||Ax - y||_2^2 + lambda||Tx - xreg||_2^2` or are we happy to use cg directly
- In order to be able to write `A^H A + lambda` as an operator do we need an additional unity operator or can this already be achieved with our current operator classes
- Would it make sense to have an (Linear)AcquisitionModel Operator which includes the functionality of "from_kdata", "recalculate_csm", "recalculate_fourier_op"... which is currently in `DirectReconstruction`. `DirectReconstruction` could then also get `acquisition_model` as input.

### LINKED ISSUE #369: Should IterativeSense have regularization?
Should Iterative sense have regularization parameters in the `__call__` ?
Or, alternatively, should this be a separate Reconstruction class, called RegularizedIterativeSENSE?

@fzimmermann89 and @ckolbPTB were in #287 for adding the parameters and keeping the name IterativeSENSE.

@koflera suggested a different class for this, as it is a different type of reconstruction.

My suggestions would be to allow in `__call__` to provide a lambda (defaults to 0) and a x_reg (default to None)
If lambda>0, it is tikonov regularization with lambda*||x||^2, if x_reg is also set, it performs lambda*||x-x_reg||^2 
The additional code in the reconstruction is minimal and would avoid having to duplicate much of the existing code.

A separate class would mainly copy the same stuff. If you want to, you could have a IterativeSense inheriting from RegularizedIterativeSense and setting lambda to 0... But having a separate class for this seems overkill imho.

Do you have any other opinions?

### PR DESCRIPTION
**Pull request by ckolbPTB**
closes #369 
closes #321 
closes #6

### CODE REVIEW

**Comment by fzimmermann89** [comment]
Allow for a linear operator in the regularization:
... + L*(Bx-x0)^2 ? With B defaulting to identity?

**Comment by ckolbPTB** [comment]
> With B defaulting to identity?

Would we then still omit an implementation of the identity operator and realise it with setting B: LinearOperator | float = 1. and making use of the OperatorTensorSum?

**Comment by fzimmermann89** [comment]
B as float does not make sense IMHO.

I would make it LinearOperator | None
with None being the identity (this also avoids creating an object as a default operator)

We can still have an explicit IdentityOperator, that can be used there (or used as a dummy linear operator)



Gesendet von Outlook für Android<https://aka.ms/AAb9ysg>
________________________________

File: src/mrpro/algorithms/reconstruction/RegularizedIterativeSENSEReconstruction.py [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+    n_iterations: int
+    """Number of CG iterations."""
+
+    regularization_data: torch.Tensor
+    """Regularisation data (i.e. prior) :math:`x_0`."""
+
+    regularization_weight: torch.Tensor
+    """Strength of the regularization :math:`L`."""
+
+    linear_op: LinearOperator | None
```
Comment (fzimmermann89) [inline src/mrpro/algorithms/reconstruction/RegularizedIterativeSENSEReconstruction.py]: regularization_op?
that it is a linear op is given by the type hint

File: src/mrpro/algorithms/reconstruction/RegularizedIterativeSENSEReconstruction.py [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+
+    def _self_adjoint_operator(self) -> LinearOperator:
+        """Create the self-adjoint operator.
+
+        Create the acquisition model as :math:`A = F S` if the CSM :math:`S` is defined otherwise :math:`A = F` with
+        the Fourier operator :math:`F`.
+
+        Create the self-adjoint operator as :math:`H = A^H W A` if the DCF is not None otherwise as :math:`H = A^H A`.
+        """
+        operator = self.fourier_op @ self.csm.as_operator() if self.csm is not None else self.fourier_op
```
Comment (fzimmermann89) [inline src/mrpro/algorithms/reconstruction/RegularizedIterativeSENSEReconstruction.py]: sugestion:
```
        csm_op = self.csm.as_operator()  if self.csm is not None else IdentityOp
        precondion_op =    self.dcf.as_operator() if self.dcf is not None else IdentityOp
        normal_operator = (csm_op@self.fourier_op).H @ precondition  @ (csm_op@self.fourier_op)
```

Comment (fzimmermann89) [reply src/mrpro/algorithms/reconstruction/RegularizedIterativeSENSEReconstruction.py]: also, i would just put it in the forward, tbh..

File: src/mrpro/algorithms/reconstruction/RegularizedIterativeSENSEReconstruction.py [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+        the Fourier operator :math:`F`.
+
+        Calculate the right-hand-side as :math:`b = A^H W y` if the DCF is not None otherwise as :math:`b = A^H y`.
+
+        Parameters
+        ----------
+        kdata
+            k-space data to reconstruct.
+        """
+        device = kdata.data.device
```
Comment (fzimmermann89) [inline src/mrpro/algorithms/reconstruction/RegularizedIterativeSENSEReconstruction.py]: remove

File: src/mrpro/algorithms/reconstruction/RegularizedIterativeSENSEReconstruction.py [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+        Parameters
+        ----------
+        kdata
+            k-space data to reconstruct.
+
+        Returns
+        -------
+            the reconstruced image.
+        """
+        device = kdata.data.device
```
Comment (fzimmermann89) [inline src/mrpro/algorithms/reconstruction/RegularizedIterativeSENSEReconstruction.py]: i would remove this.
the operator has a device and the data has to match.

Comment (fzimmermann89) [reply src/mrpro/algorithms/reconstruction/RegularizedIterativeSENSEReconstruction.py]: noise should be moved by calling reconstriction.cuda()

File: src/mrpro/algorithms/reconstruction/RegularizedIterativeSENSEReconstruction.py [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+        if self.noise is not None:
+            kdata = prewhiten_kspace(kdata, self.noise.to(device))
+
+        operator = self._self_adjoint_operator()
+        right_hand_side = self._right_hand_side(kdata)
+
+        # Add regularization
+        if not torch.all(self.regularization_weight == 0):
```
Comment (fzimmermann89) [inline src/mrpro/algorithms/reconstruction/RegularizedIterativeSENSEReconstruction.py:127]: i would set self.linear_op in the init to the identity op

**Comment by fzimmermann89** [comment]
can we jusit add the regularization example to the existing iterative sense example? these two examples seem quite redundant.

**Comment by ckolbPTB** [comment]
> can we jusit add the regularization example to the existing iterative sense example? these two examples seem quite redundant.

yes there is sume redundancy but from a teaching perspective I think it is easier to understand to build up the complexity of the reconstruction step by step.

**Comment by ckolbPTB** [comment]
@fzimmermann89 or @schuenke any more comments?

**Comment by schuenke** [comment]
> @fzimmermann89 or @schuenke any more comments?

Quickly went over the files again and I would say it looks good to me as it is.

**Comment by fzimmermann89** [comment]
I will check tonight.

Does it already use . gram ?

**Comment by fzimmermann89** [comment]
how would we include the dcf reweighting into a toeplitz form of the fourier op?

**Comment by fzimmermann89** [comment]
I would say we go ahead without the gram here for now and we discuss some approaches in a meeting.

**Comment by fzimmermann89** [comment]
How important is the weighted objective compared to only using a DCF corrected x0?
The one comparison in the Prussmann paper gives a slight advantage for >30 iterations, disadvantage for fewer.

https://arxiv.org/abs/1902.09657 ( Ong, Uecker, Lustig) also argue that the reweighting of the objective function is suboptimal due to the noise reshaping. the maximum likelihood solution would be without the dcf. they suggest proper preconditioning using different algorithms.

File: src/mrpro/algorithms/reconstruction/Reconstruction.py [hunk 6]
```diff
@@ -100,15 +100,14 @@ def direct_reconstruction(self, kdata: KData) -> IData:
...
         if self.noise is not None:
-            kdata = prewhiten_kspace(kdata, self.noise.to(device))
+            kdata = prewhiten_kspace(kdata, self.noise)
         operator = self.fourier_op
         if self.csm is not None:
             operator = operator @ self.csm.as_operator()
         if self.dcf is not None:
             operator = self.dcf.as_operator() @ operator
-        operator = operator.to(device)
+        operator = operator
```
Comment (schuenke) [inline src/mrpro/algorithms/reconstruction/Reconstruction.py]: Not needed, no?!

File: src/mrpro/algorithms/reconstruction/RegularizedIterativeSENSEReconstruction.py [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+        csm: Callable | CsmData | None = CsmData.from_idata_walsh,
+        noise: KNoise | None = None,
+        dcf: DcfData | None = None,
+        *,
+        n_iterations: int = 5,
+        regularization_data: float | torch.Tensor = 0.0,
+        regularization_weight: float | torch.Tensor,
+        regularization_op: LinearOperator | None = None,
+    ) -> None:
+        """Initialize RegularizedIterativeSENSEReconstruction.
```
Comment (schuenke) [inline src/mrpro/algorithms/reconstruction/RegularizedIterativeSENSEReconstruction.py:58]: maybe add comment here as well that for non-regularized SENSE the weight can be set to 0 or the IterativeSENSEReconstruction can be used?

**Review by schuenke — APPROVED** [review]

**Pull request merged**